### PR TITLE
Adding impi to repack

### PR DESCRIPTION
**Pull request by napetrov**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by napetrov** [comment]
@conda-forge-admin, please rerender

**Comment by napetrov** [comment]
@conda-forge-admin, please rerender

**Comment by dalcinl** [comment]
@napetrov I always found the `impi_rt` name a bit ugly. Should we really follow that pattern here? That is not the convention for other packages in this repack feedstock. Perhaps you should actually output `impi`, `impi-include`, `impi-devel` ?

**Comment by napetrov** [comment]
> @napetrov I always found the `impi_rt` name a bit ugly. Should we really follow that pattern here? That is not the convention for other packages in this repack feedstock. Perhaps you should actually output `impi`, `impi-include`, `impi-devel` ?

This feedstock is doing as is copy, so I think messing name here would case more problems. But agree on a bit strange _rt naming

**Comment by dalcinl** [comment]
> > @napetrov I always found the `impi_rt` name a bit ugly. Should we really follow that pattern here? That is not the convention for other packages in this repack feedstock. Perhaps you should actually output `impi`, `impi-include`, `impi-devel` ?
> 
> This feedstock is doing as is copy, so I think messing name here would case more problems. But agree on a bit strange _rt naming

Maybe add an additional `impi` output as a metapackage for the concrete `impi_rt` ?

**Comment by napetrov** [comment]
> > > @napetrov I always found the `impi_rt` name a bit ugly. Should we really follow that pattern here? That is not the convention for other packages in this repack feedstock. Perhaps you should actually output `impi`, `impi-include`, `impi-devel` ?
> > 
> > 
> > This feedstock is doing as is copy, so I think messing name here would case more problems. But agree on a bit strange _rt naming
> 
> Maybe add an additional `impi` output as a metapackage for the concrete `impi_rt` ?

i can pass feedback to Intel MPI team, but  first renames or additional names should be added to origin in Intel channel.  I would like to keep repack feedstock as just repack without additional functionality

**Comment by napetrov** [comment]
@isuruf @oleksandr-pavlyk  - this one is ready for merge

File: recipe/meta.yaml [hunk 1]
```diff
@@ -18,9 +18,16 @@
 {% set openmp_buildnum = "49499" %}  # [osx]
 {% set openmp_buildnum = "49496" %}  # [win]
 
+# this variable defines openmp specific version
```
Comment (dalcinl) [inline recipe/meta.yaml:21]:
```suggestion
# this variable defines impi specific version
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -18,9 +18,16 @@
...
+{% set impi_version = "2021.10.0" %}
+# this variable defines openmp specific buildnum
```
Comment (dalcinl) [inline recipe/meta.yaml:23]:
```suggestion
# this variable defines impi specific buildnum
```

**Comment by napetrov** [comment]
@dalcinl  - thanks for catching!

**Comment by dalcinl** [comment]
I know it is off-topic for this PR, but I'm wondering why the `version` and `buildnum` variables do not have a `dal_` prefix.

**Comment by napetrov** [comment]
> I know it is off-topic for this PR, but I'm wondering why the `version` and `buildnum` variables do not have a `dal_` prefix.

This would be thing to cleanup and fix(in separate PR) - but basically it's results of Conway's law. DAL team need this for pushing oneDAL repacks and other components are just get updated together with oneDAL updates. So with set of historical reasons this is how things evolved.

**Comment by dalcinl** [comment]
@napetrov Your test commands do not look right to me, they are too generic. They probably need to be different for Linux and Windows as well. Let me take a look.

**Comment by napetrov** [comment]
> @napetrov Your test commands do not look right to me, they are too generic. They probably need to be different for Linux and Windows as well. Let me take a look.

Thanks, yes this is verry welcome. There was usually limited interest in getting things checked after repack as most of validation happens for packages at Intel repo. But there is for sure set of things that can be checked to validate potential impact of repackaging

File: recipe/meta.yaml [hunk 3]
```diff
@@ -345,6 +359,61 @@ outputs:
...
+         - impi_rt/info/licenses/tpp.txt
+      description: |
+        Intel® MPI Library is a multifabric message-passing library that implements the open 
+        source MPICH specification. Use the library to create, maintain, and test advanced, 
+        complex applications that perform better on HPC clusters based on Intel® and 
+        compatible processors.
+      doc_url: https://www.intel.com/content/www/us/en/developer/tools/oneapi/mpi-library.html
+    test:
+      commands:
+        - ls -A $PREFIX/lib/*  # [unix]
```
Comment (dalcinl) [inline recipe/meta.yaml:385]:
```suggestion
        - test -f $PREFIX/bin/mpiexec  # [unix]
        - test -f $PREFIX/bin/impi_info  # [unix]
        - test -f $PREFIX/lib/libmpi.so  # [unix]
        - if not exist %LIBRARY_BIN%\\mpiexec.exe exit 1  # [win]
        - if not exist %LIBRARY_BIN%\\impi_info.exe exit 1  # [win]
        - if not exist %LIBRARY_bin%\\impi.dll exit 1  # [win]
```

File: recipe/meta.yaml [hunk 4]
```diff
@@ -345,6 +359,61 @@ outputs:
...
+         - impi-devel/info/licenses/tpp.txt
+      description: |
+        Intel® MPI Library is a multifabric message-passing library that implements the open 
+        source MPICH specification. Use the library to create, maintain, and test advanced, 
+        complex applications that perform better on HPC clusters based on Intel® and 
+        compatible processors.
+      doc_url: https://www.intel.com/content/www/us/en/developer/tools/oneapi/mpi-library.html
+    test:
+      commands:
+        - ls -A $PREFIX/lib/*  # [unix]
```
Comment (dalcinl) [inline recipe/meta.yaml:415]:
```suggestion
        - test -f $PREFIX/bin/mpicc  # [unix]
        - test -f $PREFIX/bin/mpifc  # [unix]
        - test -f $PREFIX/bin/mpicxx  # [unix]
        - test -f $PREFIX/include/mpi.h  # [unix]
        - if not exist %LIBRARY_BIN%\\mpicc.bat exit 1  # [win]
        - if not exist %LIBRARY_BIN%\\mpicl.bat exit 1  # [win]
        - if not exist %LIBRARY_BIN%\\mpifc.bat exit 1  # [win]
        - if not exist %LIBRARY_BIN%\\mpicxx.bat exit 1  # [win]
        - if not exist %LIBRARY_INC%\\mpi.h exit 1  # [win]
        - if not exist %LIBRARY_LIB%\\impi.lib exit 1  # [win]
```

**Comment by dalcinl** [comment]
I just added the improved tests as suggestions. I have not actually tried them myself as I don't have a Windows box at hand, let's see how it goes.

**Comment by dalcinl** [comment]
@napetrov Again, off-topic: Do you have direct access to Intel MPI dev/maintenance team? I've tried to communicate with Intel about issues with the Intel MPI implementation using the official community channel, but the experience was high-latency and thus a bit of a pain. There are some outstanding issues with mpi4py on Linux and Windows, a few minor details in the conda packages (e.g. missing `mpifort` compiler wrapper to make things compatible with mpich/openmpi). If you know any of the interested parties, please tell them I would like to get a more direct and straightforward interaction to work on improving the experience of my users and their users.

**Comment by napetrov** [comment]
> @napetrov Again, off-topic: Do you have direct access to Intel MPI dev/maintenance team? I've tried to communicate with Intel about issues with the Intel MPI implementation using the official community channel, but the experience was high-latency and thus a bit of a pain. There are some outstanding issues with mpi4py on Linux and Windows, a few minor details in the conda packages (e.g. missing `mpifort` compiler wrapper to make things compatible with mpich/openmpi). If you know any of the interested parties, please tell them I would like to get a more direct and straightforward interaction to work on improving the experience of my users and their users.

Done.

File: recipe/meta.yaml [hunk 5]
```diff
@@ -345,6 +359,96 @@ outputs:
...
+        compatible processors.
+      doc_url: https://www.intel.com/content/www/us/en/developer/tools/oneapi/mpi-library.html
+    test:
+      commands:
+        - test -f $PREFIX/bin/mpiexec  # [unix]
+        - test -f $PREFIX/bin/impi_info  # [unix]
+        - test -f $PREFIX/lib/libmpi.so  # [unix]
+        - if not exist %LIBRARY_BIN%\\mpiexec.exe exit 1  # [win]
+        - if not exist %LIBRARY_BIN%\\impi_info.exe exit 1  # [win]
+        - if not exist %LIBRARY_bin%\\impi.dll exit 1  # [win]
```
Comment (oleksandr-pavlyk) [inline recipe/meta.yaml:411]:
```suggestion
        - if not exist %LIBRARY_BIN%\\impi.dll exit 1  # [win]
```

**Review by oleksandr-pavlyk — APPROVED** [review]
Modulo small suggested change looks good to me. Thank you, @napetrov and @dalcinl !

**Comment by isuruf** [comment]
I prefer `intel-mpi` instead of `impi`. I had no idea of this abbreviation and that's what I searched for in anaconda.org when trying to find this package.

**Comment by dalcinl** [comment]
> I prefer `intel-mpi` instead of `impi`. I had no idea of this abbreviation and that's what I searched for in anaconda.org when trying to find this package.

Well, "impi" and "I_MPI_" are abbreviation/prefixes that intel has used forever to refer to its MPI implementation , and it in line with what other vendors have used in the past (e.g. hpmpi). We have "msmpi" in conda-forge as our own precedent of using such abbreviations.  "impi" is even used a the dll name on Windows. Any actual use of Intel MPI would be pretty much used to this abbreviation.

**Comment by dalcinl** [comment]
@napetrov @isuruf Something this not went well. The version number of the package in conda-forge is wrong, it does not match the one in the original `intel` channel.
https://anaconda.org/intel/impi-devel/files
https://anaconda.org/conda-forge/impi-devel/files

**Comment by dalcinl** [comment]
@napetrov The issue is obvious: the new outputs are missing the `version: {{ impi_version }}` entry. The `number: {{ impi_buildnum }}` entry is also missing.

**Pull request merged**